lib/portlayer/exec/base.go

```diff
@@ -248,22 +248,25 @@ func (c *containerBase) kill(ctx context.Context) error {
 	log.Infof("sending kill -%s %s", sig, c.ExecConfig.ID)
 
 	err := c.startGuestProgram(timeout, "kill", sig)
-	if err == nil {
-		log.Infof("waiting %s for %s to power off", wait, c.ExecConfig.ID)
-		err := c.vm.WaitForPowerState(timeout, types.VirtualMachinePowerStatePoweredOff)
-		if err == nil {
-			return nil // VM has powered off
-		}
-
-		if timeout.Err() != nil {
-			log.Warnf("timeout (%s) waiting for %s to power off via SIG%s", wait, c.ExecConfig.ID, sig)
-		}
+	if err == nil && timeout.Err() != nil {
+		log.Warnf("timeout (%s) waiting for %s to power off via SIG%s", wait, c.ExecConfig.ID, sig)
 	}
-
 	if err != nil {
 		log.Warnf("killing %s attempt resulted in: %s", c.ExecConfig.ID, err)
 	}
 
+	// Even if strtGuestProgram failed above, it may actually have executed.  If the container came up and then
+	// we kill it before VC gets a chance to detect the toolbox, vSphere can execute the kill but report an
+	// error indicating the guest toolbox wasn't found.  If we then try to poweroff, it may throw vSphere into
+	// an invalid transition and will need to recover.  We work around this by waiting on the power state,
+	// regardless of error from startGuestProgram.
+	// https://github.com/vmware/vic/issues/5803
+	log.Infof("waiting %s for %s to power off", wait, c.ExecConfig.ID)
+	err = c.vm.WaitForPowerState(timeout, types.VirtualMachinePowerStatePoweredOff)
```

Comment:
I guess I'm not seeing it...but how is this different than before?  If there's no `err` we will `WaitForPowerState` - isn't that what we did before?